tracepoint_perf/src/input_file.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+use core::mem;
+use core::slice;
+
+use std::fs;
+use std::io;
+use std::io::Read;
+use std::io::Seek;
+use std::vec;
+
+#[derive(Debug)]
+pub(crate) struct InputFile {
+    inner: fs::File,
+    inner_pos: u64,
+    inner_len: u64,
+}
+
+impl InputFile {
+    pub fn new(path: &str) -> io::Result<Self> {
+        let mut options = fs::OpenOptions::new();
+        options.read(true);
+
+        #[cfg(windows)]
+        {
+            use std::os::windows::fs::OpenOptionsExt;
+            const FILE_SHARE_READ: u32 = 0x00000001;
+            const FILE_SHARE_DELETE: u32 = 0x00000004;
+            options.share_mode(FILE_SHARE_READ | FILE_SHARE_DELETE);
+        }
+
+        return Ok(Self {
+            inner: options.open(path)?,
+            inner_pos: 0,
+            inner_len: 0,
+        });
+    }
+
+    pub fn len(&self) -> u64 {
+        self.inner_len
+    }
+
+    pub fn pos(&self) -> u64 {
+        self.inner_pos
+    }
+
+    pub fn update_len(&mut self) -> io::Result<()> {
+        self.inner_len = self.inner.metadata()?.len();
+        return Ok(());
+    }
+
+    pub fn seek_absolute(&mut self, new_pos: u64) -> io::Result<u64> {
+        if new_pos == self.inner_pos {
+            return Ok(new_pos);
+        } else if new_pos > self.inner_len {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "seek past end of file",
+            ));
+        } else {
+            self.inner_pos = self.inner.seek(io::SeekFrom::Start(new_pos))?;
+            return Ok(self.inner_pos);
+        }
+    }
+
+    pub fn read_exact(&mut self, buf: &mut [u8]) -> io::Result<()> {
+        self.inner.read_exact(buf)?;
+        self.inner_pos += buf.len() as u64;
+        return Ok(());
+    }
+
+    pub fn read_struct<T>(&mut self, value: &mut T) -> io::Result<()>
+    where
+        T: Copy, // Proxy for "T is a plain-old-data struct"
+    {
+        // Safety: Turning struct into slice-of-byte.
```

Comment:
The comment is *technically* true, but the actual thing being done is the opposite, right? Turning a slice-of-byte from the file into a struct?

If we're going to be deserializing a struct from a file or other byte stream, we probably should use an existing deserialization crate that will do more safety and checks for us (and will be more strict than just T: Copy).

Reply:
The structs involved are all ABI structs declared with `repr(C)`. The deserialization crate is just going to be "please declare the struct and then I'll memcpy it for you", which is what I'm doing here myself.

Yeah, technically the comment is misleading.

And yeah, T : Copy is not perfectly accurate, but it's close enough for a module-private function. It's just there to be a speed-bump for coding errors.